contracts/src/mocks/MockFunctionGateway.sol

```diff
@@ -1,4 +1,184 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.16;
 
-contract MockFunctionGateway {}
+import {VmSafe} from "forge-std/Vm.sol";
+import {stdJson} from "forge-std/StdJson.sol";
+import {IFunctionGateway} from "../interfaces/IFunctionGateway.sol";
+
+/// @title MockFunctionGateway
+/// @notice A Mock version of FunctionGateway for testing.
+/// @dev This contract is only meant to be used in tests. To use it, create a json fixture with keys "input"
+///      and "output". Call loadFixture() with the path, then this contract will automatically fulfill
+///      requests with the output.
+contract MockFunctionGateway is IFunctionGateway {
+    VmSafe private constant vm = VmSafe(address(uint160(uint256(keccak256("hevm cheat code")))));
+    uint32 public nonce;
+    bytes32 public verifiedFunctionId;
+    bytes32 public verifiedInputHash;
+    bytes public verifiedOutput;
+    bool public isCallback;
+    mapping(uint32 => bytes32) public requests;
+    mapping(bytes32 => bytes) public outputs;
+
+    function loadFixture(string memory _path) external {
+        string memory json;
+
+        try vm.readFile(_path) returns (string memory data) {
+            json = data;
+        } catch {
+            revert("MockFunctionGateway: fixture not found");
+        }
+
+        bytes memory input = stdJson.readBytes(json, "$.input");
+        bytes memory output = stdJson.readBytes(json, "$.output");
+        bytes32 inputHash = sha256(input);
+        outputs[inputHash] = output;
+    }
+
+    function requestCallback(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _context,
+        bytes4 _callbackSelector,
+        uint32 _callbackGasLimit
+    ) public payable returns (bytes32) {
+        bytes32 inputHash = sha256(_input);
+        bytes32 contextHash = keccak256(_context);
+        address callbackAddress = msg.sender;
+        bytes32 requestHash = _requestHash(
+            nonce,
+            _functionId,
+            inputHash,
+            contextHash,
+            callbackAddress,
+            _callbackSelector,
+            _callbackGasLimit
+        );
+
+        requests[nonce] = requestHash;
+        emit RequestCallback(
+            nonce,
+            _functionId,
+            _input,
+            _context,
+            callbackAddress,
+            _callbackSelector,
+            _callbackGasLimit,
+            msg.value
+        );
+        nonce++;
+
+        // If fixture has been pre-loaded, automatically fulfill and callback.
+        bytes memory output = outputs[inputHash];
+        if (output.length > 0) {
```

Comment:
I think sometimes a valid output can actually be 0 bytes--for example if your proof is just verifying that f(x) holds true if that makes sense.

Reply:
Removed